File: app/src/main/java/com/example/ta_avance/actividades/ReservasActivity.java
package com.example.ta_avance.actividades;

import android.content.Intent;
import android.content.res.Configuration;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.ta_avance.R;
import com.example.ta_avance.adapters.ReservaAdapter;
import com.example.ta_avance.dto.login.LoginRequest;
import com.example.ta_avance.dto.reserva.DtoReserva;
import com.example.ta_avance.viewmodel.ReservasViewModel;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.datepicker.CalendarConstraints;
import com.google.android.material.datepicker.CompositeDateValidator;
import com.google.android.material.datepicker.DateValidatorPointBackward;
import com.google.android.material.datepicker.DateValidatorPointForward;
import com.google.android.material.datepicker.MaterialDatePicker;
import com.google.android.material.textfield.TextInputEditText;

import java.time.DayOfWeek;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.Arrays;
import java.util.Locale;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class ReservasActivity extends AppCompatActivity {

    private ReservasViewModel viewModel;
    private RecyclerView recyclerView;
    private TextInputEditText etFechaSeleccionada;
    private Spinner spinnerEstado;
    private Button btnFiltrar;
    private TextView tituloReservas;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_reservas);

        Locale nuevoLocale = new Locale("es", "ES");
        Locale.setDefault(nuevoLocale);
        Configuration config = getResources().getConfiguration();
        config.setLocale(nuevoLocale);
        getResources().updateConfiguration(config, getResources().getDisplayMetrics());

        recyclerView = findViewById(R.id.recyclerViewReservas);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));

        spinnerEstado = findViewById(R.id.spinnerEstado);
        etFechaSeleccionada = findViewById(R.id.etFechaSeleccionada);
        btnFiltrar = findViewById(R.id.btnFiltrar);
        tituloReservas = findViewById(R.id.tituloReservas);

        String[] estados = {"CREADA", "CONFIRMADA", "REALIZADA"};
        ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.simple_spinner_item, estados);
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinnerEstado.setAdapter(adapter);

        viewModel = new ViewModelProvider(this).get(ReservasViewModel.class);

        viewModel.getReservas().observe(this, reservas -> {
            String estado = spinnerEstado.getSelectedItem().toString();
            ReservaAdapter adapterRv = new ReservaAdapter(reservas, new ReservaAdapter.OnReservaClickListener() {
                @Override
                public void onVerDetallesClick(DtoReserva reserva) {
                    mostrarPopupDetalle(reserva);
                }

                @Override
                public void onReservaRealizadaClick(DtoReserva reserva) {
                    viewModel.cambiarEstadoReserva(reserva.getReservaId(), "REALIZADA", "Reserva completada correctamente");
                }
            }, estado);
            recyclerView.setAdapter(adapterRv);
        });

        viewModel.getMensajeError().observe(this, error ->
                Toast.makeText(this, error, Toast.LENGTH_SHORT).show()
        );

        viewModel.getCambioEstadoExitoso().observe(this, exitoso -> {
            if (Boolean.TRUE.equals(exitoso)) {
                Toast.makeText(this, "Estado actualizado", Toast.LENGTH_SHORT).show();
                cargarReservas();
            }
        });

        spinnerEstado.setOnItemSelectedListener(new android.widget.AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(android.widget.AdapterView<?> parent, View view, int position, long id) {
                etFechaSeleccionada.setText("");
            }

            @Override
            public void onNothingSelected(android.widget.AdapterView<?> parent) { }
        });

        etFechaSeleccionada.setOnClickListener(v -> mostrarDatePicker());
        btnFiltrar.setOnClickListener(v -> cargarReservas());
    }

    private void mostrarDatePicker() {
        String estadoSeleccionado = spinnerEstado.getSelectedItem().toString();
        LocalDate hoy = LocalDate.now();
        MaterialDatePicker.Builder<Long> builder = MaterialDatePicker.Builder.datePicker()
                .setTitleText("Selecciona una fecha");

        if (estadoSeleccionado.equals("CREADA") || estadoSeleccionado.equals("CONFIRMADA")) {
            LocalDate finSemana = hoy.with(DayOfWeek.SUNDAY);
            long fechaMin = hoy.atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();
            long fechaMax = finSemana.atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();

            CalendarConstraints constraints = new CalendarConstraints.Builder()
                    .setStart(fechaMin)
                    .setEnd(fechaMax)
                    .setValidator(CompositeDateValidator.allOf(Arrays.asList(
                            DateValidatorPointForward.from(fechaMin),
                            DateValidatorPointBackward.before(fechaMax)
                    )))
                    .build();
            builder.setSelection(fechaMin).setCalendarConstraints(constraints);
        } else if (estadoSeleccionado.equals("REALIZADA")) {
            long fechaMax = hoy.atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli();
            CalendarConstraints constraints = new CalendarConstraints.Builder()
                    .setEnd(fechaMax)
                    .setValidator(DateValidatorPointBackward.before(fechaMax))
                    .build();
            builder.setSelection(fechaMax).setCalendarConstraints(constraints);
        }

        MaterialDatePicker<Long> datePicker = builder.build();
        datePicker.addOnPositiveButtonClickListener(selection -> {
            LocalDate fechaSeleccionada = Instant.ofEpochMilli(selection)
                    .atZone(ZoneId.systemDefault())
                    .toLocalDate();
            etFechaSeleccionada.setText(fechaSeleccionada.toString());
        });
        datePicker.show(getSupportFragmentManager(), "DATE_PICKER");
    }

    private void cargarReservas() {
        String fecha = etFechaSeleccionada.getText().toString();
        String estado = spinnerEstado.getSelectedItem().toString();

        if (estado.equals("CREADA") || estado.equals("CONFIRMADA")) {
            if (fecha.isEmpty()) {
                Toast.makeText(this, "Selecciona una fecha", Toast.LENGTH_SHORT).show();
                return;
            }
        }

        tituloReservas.setText("Reservas - " + estado);
        viewModel.cargarReservas(fecha, estado);
    }

    private void mostrarPopupDetalle(DtoReserva reserva) {
        View popupView = LayoutInflater.from(this).inflate(R.layout.popup_reserva_detalle, null);
        AlertDialog dialog = new AlertDialog.Builder(this).setView(popupView).create();

        if (dialog.getWindow() != null) {
            dialog.getWindow().setBackgroundDrawable(new ColorDrawable(android.graphics.Color.TRANSPARENT));
        }

        TextView tvUsuario = popupView.findViewById(R.id.tvDetalleUsuario);
        TextView tvBarbero = popupView.findViewById(R.id.tvDetalleBarbero);
        TextView tvHorario = popupView.findViewById(R.id.tvDetalleHorario);
        TextView tvServicio = popupView.findViewById(R.id.tvDetalleServicio);
        TextView tvFechaReserva = popupView.findViewById(R.id.tvDetalleFechaReserva);
        TextView tvFechaCreacion = popupView.findViewById(R.id.tvDetalleFechaCreacion);
        TextView tvPrecio = popupView.findViewById(R.id.tvDetallePrecio);
        TextView tvAdicionales = popupView.findViewById(R.id.tvDetalleAdicionales);
        ImageView ivComprobante = popupView.findViewById(R.id.ivComprobante);
        MaterialButton btnConfirmar = popupView.findViewById(R.id.btnConfirmarReserva);
        MaterialButton btnCancelar = popupView.findViewById(R.id.btnCancelarReserva);

        tvUsuario.setText("Usuario: " + reserva.getUsuarioNombre());
        tvBarbero.setText("Barbero: " + reserva.getBarberoNombre());
        tvHorario.setText("Horario: " + reserva.getHorarioRango());
        tvServicio.setText("Servicio: " + reserva.getServicioNombre());
        tvFechaReserva.setText("Fecha reserva: " + reserva.getFechaReserva());
        tvFechaCreacion.setText("Creado el: " + reserva.getFechaCreacion());
        tvPrecio.setText("Precio: s/" + reserva.getPrecioServicio());
        tvAdicionales.setText("Adicionales: " + reserva.getAdicionales());

        if (reserva.getUrlPago() != null && !reserva.getUrlPago().isEmpty()) {
            Glide.with(this).load(reserva.getUrlPago()).into(ivComprobante);
        } else {
            ivComprobante.setVisibility(View.GONE);
        }

        btnConfirmar.setOnClickListener(v -> {
            dialog.dismiss();
            viewModel.obtenerUsuarioPorId(reserva.getUsuarioId());
            viewModel.getUsuarioPorIdLiveData().observe(this, usuario -> {
                if (usuario != null) {
                    enviarWsp(usuario, reserva);
                    viewModel.cambiarEstadoReserva(reserva.getReservaId(), "CONFIRMADA", "Reserva confirmada por el administrador.");
                }
            });
        });

        btnCancelar.setOnClickListener(v -> {
            dialog.dismiss();
            mostrarPopupMotivo(reserva.getReservaId(), "CANCELADA");
        });

        dialog.show();
    }

    private void mostrarPopupMotivo(Long reservaId, String nuevoEstado) {
        View motivoView = LayoutInflater.from(this).inflate(R.layout.popup_motivo_cancelacion, null);
        AlertDialog motivoDialog = new AlertDialog.Builder(this).setView(motivoView).create();

        EditText etMotivoCancelacion = motivoView.findViewById(R.id.etMotivoCancelacion);
        MaterialButton btnEnviarMotivo = motivoView.findViewById(R.id.btnEnviarMotivo);

        btnEnviarMotivo.setOnClickListener(v -> {
            String motivo = etMotivoCancelacion.getText().toString().trim();
            viewModel.cambiarEstadoReserva(reservaId, nuevoEstado, motivo);
            motivoDialog.dismiss();
        });

        motivoDialog.show();
    }

    private void enviarWsp(LoginRequest usuario, DtoReserva reserva) {
        String numero = usuario.getCelular();
        String mensaje = "Hola *" + usuario.getNombre() + "*, tu reserva ha sido confirmada 🎉\n\n" +
                "📅 Fecha: *" + reserva.getFechaReserva() + "*\n" +
                "🕒 Horario: *" + reserva.getHorarioRango() + "*\n" +
                "💇 Servicio: *" + reserva.getServicioNombre() + "*\n\n" +
                "Por favor, acude puntual a tu cita. ¡Gracias por elegirnos! 💈";

        try {
            String uri = "https://wa.me/51" + numero + "?text=" + java.net.URLEncoder.encode(mensaje, "UTF-8");
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setData(android.net.Uri.parse(uri));
            startActivity(intent);
        } catch (Exception e) {
            Toast.makeText(this, "No se pudo abrir WhatsApp", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/example/ta_avance/actividades/ReservasActivity.kt
package com.example.ta_avance.actividades

import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.Spinner
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.ta_avance.R
import com.example.ta_avance.adapters.ReservaAdapter
import com.example.ta_avance.dto.login.LoginRequest
import com.example.ta_avance.dto.reserva.DtoReserva
import com.example.ta_avance.viewmodel.ReservasViewModel
import com.google.android.material.button.MaterialButton
import com.google.android.material.datepicker.CalendarConstraints
import com.google.android.material.datepicker.CompositeDateValidator
import com.google.android.material.datepicker.DateValidatorPointBackward
import com.google.android.material.datepicker.DateValidatorPointForward
import com.google.android.material.datepicker.MaterialDatePicker
import com.google.android.material.textfield.TextInputEditText
import dagger.hilt.android.AndroidEntryPoint
import java.net.URLEncoder
import java.time.DayOfWeek
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneId
import java.util.Locale

@AndroidEntryPoint
class ReservasActivity : AppCompatActivity() {

    private lateinit var viewModel: ReservasViewModel
    private lateinit var recyclerView: RecyclerView
    private lateinit var etFechaSeleccionada: TextInputEditText
    private lateinit var spinnerEstado: Spinner
    private lateinit var tituloReservas: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_reservas)

        val nuevoLocale = Locale("es", "ES")
        Locale.setDefault(nuevoLocale)
        val config = resources.configuration
        config.setLocale(nuevoLocale)
        resources.updateConfiguration(config, resources.displayMetrics)

        recyclerView = findViewById(R.id.recyclerViewReservas)
        recyclerView.layoutManager = LinearLayoutManager(this)

        spinnerEstado = findViewById(R.id.spinnerEstado)
        etFechaSeleccionada = findViewById(R.id.etFechaSeleccionada)
        tituloReservas = findViewById(R.id.tituloReservas)

        val estados = arrayOf("CREADA", "CONFIRMADA", "REALIZADA")
        spinnerEstado.adapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, estados).also {
            it.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        }

        viewModel = ViewModelProvider(this).get(ReservasViewModel::class.java)

        viewModel.reservas.observe(this) { reservas ->
            val estado = spinnerEstado.selectedItem.toString()
            recyclerView.adapter = ReservaAdapter(reservas, object : ReservaAdapter.OnReservaClickListener {
                override fun onVerDetallesClick(reserva: DtoReserva) {
                    mostrarPopupDetalle(reserva)
                }
                override fun onReservaRealizadaClick(reserva: DtoReserva) {
                    viewModel.cambiarEstadoReserva(reserva.reservaId, "REALIZADA", "Reserva completada correctamente")
                }
            }, estado)
        }

        viewModel.mensajeError.observe(this) { error ->
            Toast.makeText(this, error, Toast.LENGTH_SHORT).show()
        }

        viewModel.cambioEstadoExitoso.observe(this) { exitoso ->
            if (exitoso == true) {
                Toast.makeText(this, "Estado actualizado", Toast.LENGTH_SHORT).show()
                cargarReservas()
            }
        }

        spinnerEstado.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
                etFechaSeleccionada.setText("")
            }
            override fun onNothingSelected(parent: AdapterView<*>?) {}
        }

        etFechaSeleccionada.setOnClickListener { mostrarDatePicker() }
        findViewById<Button>(R.id.btnFiltrar).setOnClickListener { cargarReservas() }
    }

    private fun mostrarDatePicker() {
        val estadoSeleccionado = spinnerEstado.selectedItem.toString()
        val hoy = LocalDate.now()
        val builder = MaterialDatePicker.Builder.datePicker().setTitleText("Selecciona una fecha")

        when (estadoSeleccionado) {
            "CREADA", "CONFIRMADA" -> {
                val finSemana = hoy.with(DayOfWeek.SUNDAY)
                val fechaMin = hoy.atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli()
                val fechaMax = finSemana.atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli()
                builder.setSelection(fechaMin).setCalendarConstraints(
                    CalendarConstraints.Builder()
                        .setStart(fechaMin).setEnd(fechaMax)
                        .setValidator(CompositeDateValidator.allOf(listOf(
                            DateValidatorPointForward.from(fechaMin),
                            DateValidatorPointBackward.before(fechaMax)
                        ))).build()
                )
            }
            "REALIZADA" -> {
                val fechaMax = hoy.atStartOfDay(ZoneId.systemDefault()).toInstant().toEpochMilli()
                builder.setSelection(fechaMax).setCalendarConstraints(
                    CalendarConstraints.Builder()
                        .setEnd(fechaMax)
                        .setValidator(DateValidatorPointBackward.before(fechaMax))
                        .build()
                )
            }
        }

        builder.build().apply {
            addOnPositiveButtonClickListener { selection ->
                val fechaSeleccionada = Instant.ofEpochMilli(selection)
                    .atZone(ZoneId.systemDefault()).toLocalDate()
                etFechaSeleccionada.setText(fechaSeleccionada.toString())
            }
            show(supportFragmentManager, "DATE_PICKER")
        }
    }

    private fun cargarReservas() {
        val fecha = etFechaSeleccionada.text.toString()
        val estado = spinnerEstado.selectedItem.toString()

        if ((estado == "CREADA" || estado == "CONFIRMADA") && fecha.isEmpty()) {
            Toast.makeText(this, "Selecciona una fecha", Toast.LENGTH_SHORT).show()
            return
        }

        tituloReservas.text = "Reservas - $estado"
        viewModel.cargarReservas(fecha, estado)
    }

    private fun mostrarPopupDetalle(reserva: DtoReserva) {
        val popupView = LayoutInflater.from(this).inflate(R.layout.popup_reserva_detalle, null)
        val dialog = AlertDialog.Builder(this).setView(popupView).create()
        dialog.window?.setBackgroundDrawable(ColorDrawable(Color.TRANSPARENT))

        popupView.findViewById<TextView>(R.id.tvDetalleUsuario).text = "Usuario: ${reserva.usuarioNombre}"
        popupView.findViewById<TextView>(R.id.tvDetalleBarbero).text = "Barbero: ${reserva.barberoNombre}"
        popupView.findViewById<TextView>(R.id.tvDetalleHorario).text = "Horario: ${reserva.horarioRango}"
        popupView.findViewById<TextView>(R.id.tvDetalleServicio).text = "Servicio: ${reserva.servicioNombre}"
        popupView.findViewById<TextView>(R.id.tvDetalleFechaReserva).text = "Fecha reserva: ${reserva.fechaReserva}"
        popupView.findViewById<TextView>(R.id.tvDetalleFechaCreacion).text = "Creado el: ${reserva.fechaCreacion}"
        popupView.findViewById<TextView>(R.id.tvDetallePrecio).text = "Precio: s/${reserva.precioServicio}"
        popupView.findViewById<TextView>(R.id.tvDetalleAdicionales).text = "Adicionales: ${reserva.adicionales}"

        val ivComprobante = popupView.findViewById<ImageView>(R.id.ivComprobante)
        if (!reserva.urlPago.isNullOrEmpty()) {
            Glide.with(this).load(reserva.urlPago).into(ivComprobante)
        } else {
            ivComprobante.visibility = View.GONE
        }

        popupView.findViewById<MaterialButton>(R.id.btnConfirmarReserva).setOnClickListener {
            dialog.dismiss()
            viewModel.obtenerUsuarioPorId(reserva.usuarioId)
            viewModel.usuarioPorIdLiveData.observe(this) { usuario ->
                if (usuario != null) {
                    enviarWsp(usuario, reserva)
                    viewModel.cambiarEstadoReserva(reserva.reservaId, "CONFIRMADA", "Reserva confirmada por el administrador.")
                }
            }
        }

        popupView.findViewById<MaterialButton>(R.id.btnCancelarReserva).setOnClickListener {
            dialog.dismiss()
            mostrarPopupMotivo(reserva.reservaId, "CANCELADA")
        }

        dialog.show()
    }

    private fun mostrarPopupMotivo(reservaId: Long, nuevoEstado: String) {
        val motivoView = LayoutInflater.from(this).inflate(R.layout.popup_motivo_cancelacion, null)
        val motivoDialog = AlertDialog.Builder(this).setView(motivoView).create()

        val etMotivoCancelacion = motivoView.findViewById<EditText>(R.id.etMotivoCancelacion)
        motivoView.findViewById<MaterialButton>(R.id.btnEnviarMotivo).setOnClickListener {
            viewModel.cambiarEstadoReserva(reservaId, nuevoEstado, etMotivoCancelacion.text.toString().trim())
            motivoDialog.dismiss()
        }

        motivoDialog.show()
    }

    private fun enviarWsp(usuario: LoginRequest, reserva: DtoReserva) {
        val mensaje = """
            Hola *${usuario.nombre}*, tu reserva ha sido confirmada 🎉
            
            📅 Fecha: *${reserva.fechaReserva}*
            🕒 Horario: *${reserva.horarioRango}*
            💇 Servicio: *${reserva.servicioNombre}*
            
            Por favor, acude puntual a tu cita. ¡Gracias por elegirnos! 💈
        """.trimIndent()

        try {
            val uri = "https://wa.me/51${usuario.celular}?text=${URLEncoder.encode(mensaje, "UTF-8")}"
            startActivity(Intent(Intent.ACTION_VIEW).apply { data = Uri.parse(uri) })
        } catch (e: Exception) {
            Toast.makeText(this, "No se pudo abrir WhatsApp", Toast.LENGTH_SHORT).show()
        }
    }
}